File: app/lib/core/engines/recovery_engine.dart
import 'dart:math' as math;

import 'package:lifedna/shared/enums/enums.dart';

/// Last night's sleep, as the recovery engine consumes it.
class SleepInput {
  const SleepInput({
    required this.totalMinutes,
    required this.timeInBedMinutes,
    this.goalMinutes = 480,
    this.deepMinutes,
    this.remMinutes,
    this.bedtimeStdDevMinutes,
    this.wakeStdDevMinutes,
    this.nightsOfHistory = 0,
  });

  final int totalMinutes;
  final int timeInBedMinutes;
  final int goalMinutes;
  final int? deepMinutes;
  final int? remMinutes;

  /// Standard deviation of bed/wake times across the last 14 nights. Null when
  /// there is not yet enough history — the consistency component is then simply
  /// unavailable rather than guessed.
  final double? bedtimeStdDevMinutes;
  final double? wakeStdDevMinutes;
  final int nightsOfHistory;
}

class TrainingInput {
  const TrainingInput({
    required this.acwr,
    required this.yesterdayLoad,
    required this.meanDailyLoad28d,
    this.yesterdaySessionRpe,
    this.daysSinceLastSession = 0,
  });

  /// Null when there is not enough history for a meaningful ratio.
  final double? acwr;
  final double yesterdayLoad;
  final double meanDailyLoad28d;
  final int? yesterdaySessionRpe;
  final int daysSinceLastSession;
}

class ActivityInput {
  const ActivityInput({
    required this.steps,
    required this.stepGoal,
    required this.activeMinutes,
    this.trainedYesterday = false,
  });

  final int steps;
  final int stepGoal;
  final int activeMinutes;
  final bool trainedYesterday;
}

/// Optional physiological signals. Present only when a wearable supplies them
/// and at least 14 days of baseline exist.
class PhysiologyInput {
  const PhysiologyInput({
    this.restingHrBpm,
    this.baselineRestingHrBpm,
    this.hrvMs,
    this.baselineHrvMs,
  });

  final int? restingHrBpm;
  final int? baselineRestingHrBpm;
  final double? hrvMs;
  final double? baselineHrvMs;

  bool get hasRestingHr => restingHrBpm != null && baselineRestingHrBpm != null;
  bool get hasHrv =>
      hrvMs != null && baselineHrvMs != null && baselineHrvMs! > 0;
}

/// One weighted component of the composite score, with everything needed to
/// explain it. Every score in LifeDNA is decomposable — see docs/12 §2.
class RecoveryComponent {
  const RecoveryComponent({
    required this.name,
    required this.score,
    required this.weight,
    required this.available,
  });

  final String name;
  final int score;
  final double weight;
  final bool available;

  double get contribution =>
      available ? (score * weight * 10).roundToDouble() / 10 : 0;
}

class RecoveryResult {
  const RecoveryResult({
    required this.recoveryScore,
    required this.band,
    required this.readinessScore,
    required this.components,
    required this.sleepScore,
    required this.missingInputs,
    required this.action,
    required this.detail,
    required this.engineVersion,
  });

  /// Null when fewer than two domains were available.
  final int? recoveryScore;
  final RecoveryBand band;
  final int? readinessScore;
  final List<RecoveryComponent> components;
  final int? sleepScore;
  final List<String> missingInputs;
  final TrainingAction? action;
  final String detail;
  final String engineVersion;

  bool get isSufficient => recoveryScore != null;
}

/// The Recovery Engine.
///
/// Specification: docs/12-recovery-engine.md. This is the client mirror of the
/// authoritative server implementation; both are validated against the shared
/// fixtures in `test/fixtures/engines/recovery/`. The mirror exists so the UI
/// can recompute instantly when the user edits an input, and always agree with
/// what the server will later store.
///
/// It answers exactly one question: **how hard can I train today?**
abstract final class RecoveryEngine {
  static const String version = 'recovery-1.2.0';

  // Default weights. Overridable server-side via Remote Config (RECV-02).
  static const double wSleep = 0.40;
  static const double wTraining = 0.40;
  static const double wActivity = 0.20;

  /// Fewer than two available domains produces no score at all, rather than a
  /// confident-looking guess (RECV-07).
  static const int minimumComponents = 2;

  static RecoveryResult compute({
    SleepInput? sleep,
    TrainingInput? training,
    ActivityInput? activity,
    PhysiologyInput physiology = const PhysiologyInput(),
    PlannedSession? plannedSession,
  }) {
    final missing = <String>[];

    final sleepScore = sleep == null ? null : _sleepScore(sleep);
    final trainingScore = training == null ? null : _trainingScore(training);
    final activityScore = activity == null ? null : _activityScore(activity);

    if (sleepScore == null) missing.add('sleep');
    if (trainingScore == null) missing.add('training');
    if (activityScore == null) missing.add('activity');

    final availableCount = [
      sleepScore,
      trainingScore,
      activityScore,
    ].where((s) => s != null).length;

    final components = <RecoveryComponent>[];

    if (availableCount < minimumComponents) {
      return RecoveryResult(
        recoveryScore: null,
        band: RecoveryBand.insufficientData,
        readinessScore: null,
        components: components,
        sleepScore: sleepScore,
        missingInputs: missing,
        action: null,
        detail: _insufficientDataMessage(missing),
        engineVersion: version,
      );
    }

    // Renormalize the weights over whatever is actually available, so a missing
    // domain does not silently drag the score toward zero.
    var weightTotal = 0.0;
    if (sleepScore != null) weightTotal += wSleep;
    if (trainingScore != null) weightTotal += wTraining;
    if (activityScore != null) weightTotal += wActivity;

    final normSleep = sleepScore != null ? wSleep / weightTotal : 0.0;
    final normTraining = trainingScore != null ? wTraining / weightTotal : 0.0;
    final normActivity = activityScore != null ? wActivity / weightTotal : 0.0;

    components
      ..add(
        RecoveryComponent(
          name: 'sleep',
          score: sleepScore ?? 0,
          weight: _round2(normSleep),
          available: sleepScore != null,
        ),
      )
      ..add(
        RecoveryComponent(
          name: 'training',
          score: trainingScore ?? 0,
          weight: _round2(normTraining),
          available: trainingScore != null,
        ),
      )
      ..add(
        RecoveryComponent(
          name: 'activity',
          score: activityScore ?? 0,
          weight: _round2(normActivity),
          available: activityScore != null,
        ),
      );

    var composite =
        (sleepScore ?? 0) * normSleep +
        (trainingScore ?? 0) * normTraining +
        (activityScore ?? 0) * normActivity;

    composite += _physiologyAdjustment(physiology);

    final recovery = composite.clamp(0.0, 100.0).round();
    final band = RecoveryBand.forScore(recovery);

    final readiness = _readiness(
      recovery: recovery,
      planned: plannedSession,
      meanDailyLoad28d: training?.meanDailyLoad28d ?? 0,
    );

    final action = _action(readiness ?? recovery, training?.acwr);

    return RecoveryResult(
      recoveryScore: recovery,
      band: band,
      readinessScore: readiness,
      components: components,
      sleepScore: sleepScore,
      missingInputs: missing,
      action: action,
      detail: _detail(
        action: action,
        recovery: recovery,
        readiness: readiness,
        sleepScore: sleepScore,
        acwr: training?.acwr,
      ),
      engineVersion: version,
    );
  }

  // ---------------------------------------------------------------- sleep ---

  /// `0.40·duration + 0.25·consistency + 0.20·efficiency + 0.15·stages`,
  /// renormalized over whichever sub-components are available.
  static int _sleepScore(SleepInput s) {
    final duration = _sleepDuration(s);
    final consistency = _sleepConsistency(s);
    final efficiency = _sleepEfficiency(s);
    final stages = _sleepStages(s);

    var sum = 0.0;
    var weight = 0.0;

    sum += duration * 0.40;
    weight += 0.40;

    if (consistency != null) {
      sum += consistency * 0.25;
      weight += 0.25;
    }
    if (efficiency != null) {
      sum += efficiency * 0.20;
      weight += 0.20;
    }
    if (stages != null) {
      sum += stages * 0.15;
      weight += 0.15;
    }

    return (sum / weight).clamp(0.0, 100.0).round();
  }

  /// Under-sleeping is penalized ~2.2× harder than over-sleeping, because the
  /// physiological cost is asymmetric. Long sleep is mildly penalized (it
  /// correlates with poor quality or illness) but never below 60.
  static double _sleepDuration(SleepInput s) {
    if (s.goalMinutes <= 0) return 0;
    final ratio = s.totalMinutes / s.goalMinutes;
    if (ratio >= 0.95 && ratio <= 1.15) return 100;
    if (ratio < 0.95) {
      return math.max(0, 100 - (0.95 - ratio) * 220);
    }
    return math.max(60, 100 - (ratio - 1.15) * 100);
  }

  /// Schedule stability. σ ≤ 20 min scores 100.
  static double? _sleepConsistency(SleepInput s) {
    if (s.nightsOfHistory < 7) return null;
    final bed = s.bedtimeStdDevMinutes;
    final wake = s.wakeStdDevMinutes;
    if (bed == null || wake == null) return null;
    final sigma = (bed + wake) / 2;
    return (100 - (sigma - 20) * 1.6).clamp(0.0, 100.0).toDouble();
  }

  static double? _sleepEfficiency(SleepInput s) {
    if (s.timeInBedMinutes <= 0) return null;
    final pct = s.totalMinutes / s.timeInBedMinutes * 100;
    if (pct >= 90) return 100;
    if (pct >= 60) return (pct - 60) * 3.33;
    return 0;
  }

  /// Ideal bands: deep 13–23 %, REM 20–25 %.
  static double? _sleepStages(SleepInput s) {
    final deep = s.deepMinutes;
    final rem = s.remMinutes;
    if (deep == null || rem == null || s.totalMinutes <= 0) return null;

    final deepPct = deep / s.totalMinutes * 100;
    final remPct = rem / s.totalMinutes * 100;

    final deepScore = (100 - (deepPct - 18).abs() * 5).clamp(0.0, 100.0);
    final remScore = (100 - (remPct - 22.5).abs() * 5).clamp(0.0, 100.0);
    return ((deepScore + remScore) / 2).toDouble();
  }

  // ------------------------------------------------------------- training ---

  /// A plateau, not a peak: a wide productive band with penalties on both sides.
  static int _trainingScore(TrainingInput t) {
    final acwr = t.acwr;

    // Without enough history the ratio is meaningless, so return a neutral
    // score rather than penalizing a new user for having no chronic load.
    var base = 75.0;
    if (acwr != null) {
      if (acwr < 0.60) {
        base = 70;
      } else if (acwr < 0.80) {
        base = 85;
      } else if (acwr < 1.30) {
        base = 100;
      } else if (acwr < 1.50) {
        base = 75;
      } else if (acwr < 1.80) {
        base = 50;
      } else {
        base = 25;
      }
    }

    var adjustment = 0.0;

    // A single unusually heavy session yesterday still costs today, even when
    // the ratio looks healthy.
    if (t.meanDailyLoad28d > 0 && t.yesterdayLoad > 1.5 * t.meanDailyLoad28d) {
      adjustment -= 15;
    }
    if ((t.yesterdaySessionRpe ?? 0) >= 9) {
      adjustment -= 8;
    }
    if (t.daysSinceLastSession >= 2 && (acwr == null || acwr >= 0.8)) {
      adjustment += 10;
    }

    return (base + adjustment).clamp(0.0, 100.0).round();
  }

  // ------------------------------------------------------------- activity ---

  static int _activityScore(ActivityInput a) {
    if (a.stepGoal <= 0) return 0;
    final stepRatio = a.steps / a.stepGoal;

    final double stepScore;
    if (stepRatio >= 1.0) {
      stepScore = 100;
    } else if (stepRatio >= 0.5) {
      stepScore = stepRatio * 100;
    } else {
      stepScore = stepRatio * 80;
    }

    final activeScore = (a.activeMinutes / 45 * 100)
        .clamp(0.0, 100.0)
        .toDouble();
    var score = 0.6 * stepScore + 0.4 * activeScore;

    // A genuinely sedentary day is a recovery signal in its own right.
    if (a.steps < 3000 && !a.trainedYesterday) score -= 10;

    return score.clamp(0.0, 100.0).round();
  }

  // ---------------------------------------------------------- physiology ---

  /// Capped at ±10 points in total so a noisy wearable reading can never
  /// dominate the score.
  static double _physiologyAdjustment(PhysiologyInput p) {
    var adjustment = 0.0;

    if (p.hasRestingHr) {
      final delta = p.restingHrBpm! - p.baselineRestingHrBpm!;
      if (delta <= -3) {
        adjustment += 3;
      } else if (delta >= 3 && delta <= 5) {
        adjustment -= 4;
      } else if (delta > 5) {
        adjustment -= 8;
      }
    }

    if (p.hasHrv) {
      final deltaPct = (p.hrvMs! - p.baselineHrvMs!) / p.baselineHrvMs! * 100;
      if (deltaPct >= 10) {
        adjustment += 4;
      } else if (deltaPct <= -20) {
        adjustment -= 9;
      } else if (deltaPct <= -10) {
        adjustment -= 5;
      }
    }

    return adjustment.clamp(-10.0, 10.0).toDouble();
  }

  // ----------------------------------------------------------- readiness ---

  /// Recovery is backward-looking. Readiness asks the forward question: how
  /// recovered am I *relative to what today actually demands*?
  static int? _readiness({
    required int recovery,
    required PlannedSession? planned,
    required double meanDailyLoad28d,
  }) {
    if (planned == null || meanDailyLoad28d <= 0) return recovery;
    final plannedLoad = planned.rpe * planned.durationMinutes;
    final demand = (plannedLoad / meanDailyLoad28d).clamp(0.0, 2.5);
    return (recovery - (demand - 1.0) * 18).clamp(0.0, 100.0).round();
  }

  static TrainingAction _action(int readiness, double? acwr) {
    if (readiness < 40 || (acwr != null && acwr >= 1.8)) {
      return TrainingAction.rest;
    }
    if (readiness < 65) return TrainingAction.reduce;
    if (readiness >= 80 && (acwr == null || acwr < 1.3)) {
      return TrainingAction.push;
    }
    return TrainingAction.proceed;
  }

  // --------------------------------------------------------------- copy ----

  static String _detail({
    required TrainingAction action,
    required int recovery,
    required int? readiness,
    required int? sleepScore,
    required double? acwr,
  }) {
    final ratio = acwr?.toStringAsFixed(2);
    return switch (action) {
      TrainingAction.push =>
        'Recovery $recovery with a balanced load ratio'
            '${ratio == null ? '' : ' of $ratio'}. '
            'This is the day to attempt a record.',
      TrainingAction.proceed =>
        readiness != null && readiness < recovery
            ? 'Recovery is $recovery, but today’s session is heavier than your '
                  'average, so readiness lands at $readiness. Hit your target reps; '
                  'save the record attempt for a lighter day.'
            : 'Recovery $recovery'
                  '${ratio == null ? '' : ' and a load ratio of $ratio'}. '
                  'Train as planned.',
      TrainingAction.reduce =>
        'Recovery is $recovery${sleepScore == null ? '' : ' with a sleep score '
                      'of $sleepScore'}. Keep the compound work and drop the last set of '
            'each accessory.',
      TrainingAction.rest =>
        'Recovery is $recovery'
            '${ratio == null ? '' : ' and your load ratio is $ratio'}. '
            'Training hard today costs more than it earns.',
    };
  }

  /// Names exactly what is missing. "Insufficient data" without saying what
  /// would fix it is a dead end, and the flow rules forbid dead ends.
  static String _insufficientDataMessage(List<String> missing) {
    if (missing.contains('sleep')) {
      return 'Recovery needs sleep data from at least 2 of the last 3 nights.';
    }
    final names = missing.map(_domainLabel).join(' and ');
    return 'Recovery needs at least two data sources. Still missing: $names.';
  }

  static String _domainLabel(String domain) => switch (domain) {
    'sleep' => 'sleep',
    'training' => 'training history',
    'activity' => 'daily activity',
    _ => domain,
  };

  static double _round2(double v) => (v * 100).roundToDouble() / 100;
}

/// Today's planned session, used to convert recovery into readiness.
class PlannedSession {
  const PlannedSession({required this.rpe, required this.durationMinutes});
  final int rpe;
  final int durationMinutes;
}

File: functions/src/engines/recovery/recoveryEngine.ts
/**
 * The Recovery Engine — the authoritative implementation.
 *
 * Mirrored on the client at `app/lib/core/engines/recovery_engine.dart`; both
 * run the shared fixtures in `test/fixtures/engines/recovery/`.
 *
 * Specification: docs/12-recovery-engine.md.
 *
 * It answers exactly one question: how hard can I train today?
 */

import { roundHalfAwayFromZero } from '../../lib/rounding';

export type RecoveryBand = 'low' | 'moderate' | 'high' | 'insufficient_data';
export type TrainingAction = 'push' | 'proceed' | 'reduce' | 'rest';

export interface SleepInput {
  totalMinutes: number;
  timeInBedMinutes: number;
  goalMinutes?: number;
  deepMinutes?: number | null;
  remMinutes?: number | null;
  bedtimeStdDevMinutes?: number | null;
  wakeStdDevMinutes?: number | null;
  nightsOfHistory?: number;
}

export interface TrainingInput {
  acwr: number | null;
  yesterdayLoad: number;
  meanDailyLoad28d: number;
  yesterdaySessionRpe?: number | null;
  daysSinceLastSession?: number;
}

export interface ActivityInput {
  steps: number;
  stepGoal: number;
  activeMinutes: number;
  trainedYesterday?: boolean;
}

export interface PhysiologyInput {
  restingHrBpm?: number | null;
  baselineRestingHrBpm?: number | null;
  hrvMs?: number | null;
  baselineHrvMs?: number | null;
}

export interface PlannedSession {
  rpe: number;
  durationMinutes: number;
}

export interface RecoveryComponent {
  name: 'sleep' | 'training' | 'activity';
  score: number;
  weight: number;
  contribution: number;
  available: boolean;
}

export interface RecoveryResult {
  recoveryScore: number | null;
  band: RecoveryBand;
  readinessScore: number | null;
  sleepScore: number | null;
  components: RecoveryComponent[];
  missingInputs: string[];
  action: TrainingAction | null;
  detail: string;
  engineVersion: string;
}

export const RECOVERY_ENGINE_VERSION = 'recovery-1.2.0';

/** Default weights. Overridable via Remote Config (RECV-02). */
export const W_SLEEP = 0.4;
export const W_TRAINING = 0.4;
export const W_ACTIVITY = 0.2;

/**
 * Fewer than two available domains produces NO score, rather than a
 * confident-looking guess (RECV-07).
 */
export const MINIMUM_COMPONENTS = 2;

export function computeRecovery(args: {
  sleep?: SleepInput | null;
  training?: TrainingInput | null;
  activity?: ActivityInput | null;
  physiology?: PhysiologyInput;
  plannedSession?: PlannedSession | null;
}): RecoveryResult {
  const { sleep, training, activity, physiology = {}, plannedSession } = args;

  const missing: string[] = [];
  const sleepScore = sleep ? computeSleepScore(sleep) : null;
  const trainingScore = training ? computeTrainingScore(training) : null;
  const activityScore = activity ? computeActivityScore(activity) : null;

  if (sleepScore === null) missing.push('sleep');
  if (trainingScore === null) missing.push('training');
  if (activityScore === null) missing.push('activity');

  const availableCount = [sleepScore, trainingScore, activityScore].filter(
    (s) => s !== null,
  ).length;

  if (availableCount < MINIMUM_COMPONENTS) {
    return {
      recoveryScore: null,
      band: 'insufficient_data',
      readinessScore: null,
      sleepScore,
      components: [],
      missingInputs: missing,
      action: null,
      detail: insufficientDataMessage(missing),
      engineVersion: RECOVERY_ENGINE_VERSION,
    };
  }

  // Renormalize over what is actually available, so a missing domain does not
  // silently drag the score toward zero.
  let weightTotal = 0;
  if (sleepScore !== null) weightTotal += W_SLEEP;
  if (trainingScore !== null) weightTotal += W_TRAINING;
  if (activityScore !== null) weightTotal += W_ACTIVITY;

  const normSleep = sleepScore !== null ? W_SLEEP / weightTotal : 0;
  const normTraining = trainingScore !== null ? W_TRAINING / weightTotal : 0;
  const normActivity = activityScore !== null ? W_ACTIVITY / weightTotal : 0;

  const components: RecoveryComponent[] = [
    component('sleep', sleepScore, normSleep),
    component('training', trainingScore, normTraining),
    component('activity', activityScore, normActivity),
  ];

  let composite =
    (sleepScore ?? 0) * normSleep +
    (trainingScore ?? 0) * normTraining +
    (activityScore ?? 0) * normActivity;

  composite += physiologyAdjustment(physiology);

  const recovery = clamp(roundHalfAwayFromZero(composite), 0, 100);
  const band = bandFor(recovery);
  const readiness = computeReadiness(
    recovery,
    plannedSession ?? null,
    training?.meanDailyLoad28d ?? 0,
  );
  const action = actionFor(readiness ?? recovery, training?.acwr ?? null);

  return {
    recoveryScore: recovery,
    band,
    readinessScore: readiness,
    sleepScore,
    components,
    missingInputs: missing,
    action,
    detail: detailFor(action, recovery, readiness, sleepScore, training?.acwr ?? null),
    engineVersion: RECOVERY_ENGINE_VERSION,
  };
}

// ------------------------------------------------------------------- sleep --

export function computeSleepScore(s: SleepInput): number {
  const duration = sleepDuration(s);
  const consistency = sleepConsistency(s);
  const efficiency = sleepEfficiency(s);
  const stages = sleepStages(s);

  let sum = duration * 0.4;
  let weight = 0.4;

  if (consistency !== null) {
    sum += consistency * 0.25;
    weight += 0.25;
  }
  if (efficiency !== null) {
    sum += efficiency * 0.2;
    weight += 0.2;
  }
  if (stages !== null) {
    sum += stages * 0.15;
    weight += 0.15;
  }

  return clamp(roundHalfAwayFromZero(sum / weight), 0, 100);
}

/**
 * Under-sleeping is penalized ~2.2x harder than over-sleeping, because the
 * physiological cost is asymmetric. Long sleep is mildly penalized (it
 * correlates with poor quality or illness) but never below 60.
 */
function sleepDuration(s: SleepInput): number {
  const goal = s.goalMinutes ?? 480;
  if (goal <= 0) return 0;
  const ratio = s.totalMinutes / goal;
  if (ratio >= 0.95 && ratio <= 1.15) return 100;
  if (ratio < 0.95) return Math.max(0, 100 - (0.95 - ratio) * 220);
  return Math.max(60, 100 - (ratio - 1.15) * 100);
}

function sleepConsistency(s: SleepInput): number | null {
  if ((s.nightsOfHistory ?? 0) < 7) return null;
  const bed = s.bedtimeStdDevMinutes;
  const wake = s.wakeStdDevMinutes;
  if (bed === null || bed === undefined || wake === null || wake === undefined) {
    return null;
  }
  const sigma = (bed + wake) / 2;
  return clamp(100 - (sigma - 20) * 1.6, 0, 100);
}

function sleepEfficiency(s: SleepInput): number | null {
  if (s.timeInBedMinutes <= 0) return null;
  const pct = (s.totalMinutes / s.timeInBedMinutes) * 100;
  if (pct >= 90) return 100;
  if (pct >= 60) return (pct - 60) * 3.33;
  return 0;
}

/** Ideal bands: deep 13–23 %, REM 20–25 %. */
function sleepStages(s: SleepInput): number | null {
  const deep = s.deepMinutes;
  const rem = s.remMinutes;
  if (deep === null || deep === undefined) return null;
  if (rem === null || rem === undefined) return null;
  if (s.totalMinutes <= 0) return null;

  const deepPct = (deep / s.totalMinutes) * 100;
  const remPct = (rem / s.totalMinutes) * 100;
  const deepScore = clamp(100 - Math.abs(deepPct - 18) * 5, 0, 100);
  const remScore = clamp(100 - Math.abs(remPct - 22.5) * 5, 0, 100);
  return (deepScore + remScore) / 2;
}

// ---------------------------------------------------------------- training --

/** A plateau, not a peak: a wide productive band with penalties on both sides. */
export function computeTrainingScore(t: TrainingInput): number {
  const acwr = t.acwr;

  // Without enough history the ratio is meaningless, so return a neutral
  // score rather than penalizing a new user for having no chronic load.
  let base = 75;
  if (acwr !== null) {
    if (acwr < 0.6) base = 70;
    else if (acwr < 0.8) base = 85;
    else if (acwr < 1.3) base = 100;
    else if (acwr < 1.5) base = 75;
    else if (acwr < 1.8) base = 50;
    else base = 25;
  }

  let adjustment = 0;
  if (t.meanDailyLoad28d > 0 && t.yesterdayLoad > 1.5 * t.meanDailyLoad28d) {
    adjustment -= 15;
  }
  if ((t.yesterdaySessionRpe ?? 0) >= 9) adjustment -= 8;
  if ((t.daysSinceLastSession ?? 0) >= 2 && (acwr === null || acwr >= 0.8)) {
    adjustment += 10;
  }

  return clamp(roundHalfAwayFromZero(base + adjustment), 0, 100);
}

// ---------------------------------------------------------------- activity --

export function computeActivityScore(a: ActivityInput): number {
  if (a.stepGoal <= 0) return 0;
  const stepRatio = a.steps / a.stepGoal;

  let stepScore: number;
  if (stepRatio >= 1) stepScore = 100;
  else if (stepRatio >= 0.5) stepScore = stepRatio * 100;
  else stepScore = stepRatio * 80;

  const activeScore = clamp((a.activeMinutes / 45) * 100, 0, 100);
  let score = 0.6 * stepScore + 0.4 * activeScore;

  // A genuinely sedentary day is a recovery signal in its own right.
  if (a.steps < 3000 && !(a.trainedYesterday ?? false)) score -= 10;

  return clamp(roundHalfAwayFromZero(score), 0, 100);
}

// -------------------------------------------------------------- physiology --

/** Capped at ±10 so a noisy wearable reading can never dominate the score. */
function physiologyAdjustment(p: PhysiologyInput): number {
  let adjustment = 0;

  const rhr = p.restingHrBpm;
  const rhrBase = p.baselineRestingHrBpm;
  if (rhr !== null && rhr !== undefined && rhrBase !== null && rhrBase !== undefined) {
    const delta = rhr - rhrBase;
    if (delta <= -3) adjustment += 3;
    else if (delta >= 3 && delta <= 5) adjustment -= 4;
    else if (delta > 5) adjustment -= 8;
  }

  const hrv = p.hrvMs;
  const hrvBase = p.baselineHrvMs;
  if (hrv !== null && hrv !== undefined && hrvBase !== null && hrvBase !== undefined && hrvBase > 0) {
    const deltaPct = ((hrv - hrvBase) / hrvBase) * 100;
    if (deltaPct >= 10) adjustment += 4;
    else if (deltaPct <= -20) adjustment -= 9;
    else if (deltaPct <= -10) adjustment -= 5;
  }

  return clamp(adjustment, -10, 10);
}

// --------------------------------------------------------------- readiness --

/**
 * Recovery is backward-looking. Readiness asks the forward question: how
 * recovered am I relative to what today actually demands?
 */
function computeReadiness(
  recovery: number,
  planned: PlannedSession | null,
  meanDailyLoad28d: number,
): number | null {
  if (planned === null || meanDailyLoad28d <= 0) return recovery;
  const plannedLoad = planned.rpe * planned.durationMinutes;
  const demand = clamp(plannedLoad / meanDailyLoad28d, 0, 2.5);
  return clamp(roundHalfAwayFromZero(recovery - (demand - 1) * 18), 0, 100);
}

function actionFor(readiness: number, acwr: number | null): TrainingAction {
  if (readiness < 40 || (acwr !== null && acwr >= 1.8)) return 'rest';
  if (readiness < 65) return 'reduce';
  if (readiness >= 80 && (acwr === null || acwr < 1.3)) return 'push';
  return 'proceed';
}

export function bandFor(score: number): RecoveryBand {
  if (score <= 33) return 'low';
  if (score <= 66) return 'moderate';
  return 'high';
}

// -------------------------------------------------------------------- copy --

function detailFor(
  action: TrainingAction,
  recovery: number,
  readiness: number | null,
  sleepScore: number | null,
  acwr: number | null,
): string {
  const ratio = acwr === null ? null : acwr.toFixed(2);
  switch (action) {
    case 'push':
      return `Recovery ${recovery} with a balanced load ratio${
        ratio === null ? '' : ` of ${ratio}`
      }. This is the day to attempt a record.`;
    case 'proceed':
      return readiness !== null && readiness < recovery
        ? `Recovery is ${recovery}, but today's session is heavier than your average, so readiness lands at ${readiness}. Hit your target reps; save the record attempt for a lighter day.`
        : `Recovery ${recovery}${
            ratio === null ? '' : ` and a load ratio of ${ratio}`
          }. Train as planned.`;
    case 'reduce':
      return `Recovery is ${recovery}${
        sleepScore === null ? '' : ` with a sleep score of ${sleepScore}`
      }. Keep the compound work and drop the last set of each accessory.`;
    case 'rest':
      return `Recovery is ${recovery}${
        ratio === null ? '' : ` and your load ratio is ${ratio}`
      }. Training hard today costs more than it earns.`;
  }
}

/**
 * Names exactly what is missing. "Insufficient data" without saying what would
 * fix it is a dead end, and the flow rules forbid dead ends.
 */
function insufficientDataMessage(missing: string[]): string {
  if (missing.includes('sleep')) {
    return 'Recovery needs sleep data from at least 2 of the last 3 nights.';
  }
  const names = missing.map(domainLabel).join(' and ');
  return `Recovery needs at least two data sources. Still missing: ${names}.`;
}

function domainLabel(domain: string): string {
  switch (domain) {
    case 'sleep':
      return 'sleep';
    case 'training':
      return 'training history';
    case 'activity':
      return 'daily activity';
    default:
      return domain;
  }
}

function component(
  name: RecoveryComponent['name'],
  score: number | null,
  weight: number,
): RecoveryComponent {
  const available = score !== null;
  const roundedWeight = roundHalfAwayFromZero(weight * 100) / 100;
  return {
    name,
    score: score ?? 0,
    weight: roundedWeight,
    contribution: available
      ? roundHalfAwayFromZero((score ?? 0) * roundedWeight * 10) / 10
      : 0,
    available,
  };
}

function clamp(value: number, lo: number, hi: number): number {
  return Math.min(hi, Math.max(lo, value));
}
